### DISCUSSION TITLE
How to make Clippy fail on warnings, but keep warnings printed as warnings?

### DISCUSSION THREAD

**Discussion by laniakea64** [rust-lang/rust-clippy#11942] [Q&A] [answered]
Relatively new to Rust and Clippy.  I'm trying to create a [justfile](https://just.systems/) recipe that first runs Clippy (`cargo clippy --all-targets --all-features`), then if Clippy gives no errors or warnings, run `cargo fmt --all -- --check`.  But if there were only warnings, Clippy still exits with status 0, so the justfile recipe does not stop running there and goes on to the formatting check.

One way to make Clippy exit with non-zero status on warnings is to append `-- -Dwarnings` to the Clippy command.  But this also makes warnings printed as errors, and with me not being an experienced Rustacean, seeing red "error" for warnings is confusing.

How to make Clippy exit with non-zero status when there were warnings, but keep warnings printed as warnings?
If this is not possible, could such option please be added?

Thanks!

**Comment by flip1995** [comment]
>  If this is not possible, could such option please be added?

No this is not possible and that's by design.

In Rust, warnings are considered (almost) as problematic as errors. So CI systems are usually set up to treat warnings as errors (`-Dwarnings`).

The reason, why warnings will give exit status `0` is that for local prototyping/development/experimentation you don't really care if there is an unused variable temporarily, so you don't want to fail compilation on warnings. That is **locally**, i.e. in your `Justfile`.

Also read about the [`#![deny(warnings)]`](https://rust-unofficial.github.io/patterns/anti_patterns/deny-warnings.html) anti-pattern.

TL;DR: What you want to do is to keep warnings as warnings locally and in CI set `RUSTFLAGS="-Dwarnings"`

**Reply by laniakea64** [reply]
Thanks flip1995 for the answer.  I'm confused what failing compilation has to do with this?  `cargo build` and `cargo run` are handled in other justfile recipes, and you are right that I don't want those to stop/fail when there are only warnings.

**Reply by flip1995** [reply, marked as answer]
If you specify `deny(warnings)` or `-Dwarnings` you tell the Rust compiler (or Clippy) that you want to fail compilation on warnings. That's exactly what this flag (or attribute) means.

> `cargo build` and `cargo run` are handled in other justfile recipes

In that case, you might want to set `-Dwarnings` in only the `Justfile` where you want to fail on warnings and not in those where you don't.

**Comment by laniakea64** [comment]
Looks like this question stemmed from a misconception.  To make sure I understood flip1995's answers correctly:

- In rustc output, "error" and "warning" are **not** indicating different levels of severity.  The difference between "error" and "warning" is whether an issue caused compilation to fail.
- Clippy works by providing lints for rustc to emit during compilation.  Clippy doesn't directly emit warnings/errors itself.  Appending `-- -Dwarnings` to the Clippy command is nothing more than passing `-Dwarnings` to rustc.
- Because it would not make sense for Clippy or rustc to exit with non-zero status on a successful compilation, the solution to the original issue is to append `-- -Dwarnings` only to the Clippy command, and understand what "error" and "warning" mean in Rust.

Do I have it right?

**Reply by flip1995** [reply]
> * In rustc output, "error" and "warning" are **not** indicating different levels of severity

They indicate different levels of severity in `rustc`. I'm saying that the Rust community considers warnings in production code just as severe as errors. So you will see `-Dwarnings` in most CIs

> * to append `-- -Dwarnings` only to the Clippy command

Not quite. You should globally set `RUSTFLAGS="-Dwarnings"` _**in CI**_. Locally, you shouldn't specify `-Dwarnings` (or similar) anywhere. See the [`#![deny(warnings)]`](https://rust-unofficial.github.io/patterns/anti_patterns/deny-warnings.html) anti-pattern.

---

Point 2 is about right.

**Reply by laniakea64** [reply]
Appreciate the clarification, I get it now.  Thanks for your help flip1995! :smiley:
